lib/Driver/Compilation.cpp

```diff
@@ -421,14 +424,31 @@ int Compilation::performJobsImpl() {
   }
 
   int Result = EXIT_SUCCESS;
+  llvm::TimerGroup DriverTimerGroup("Driver Time Compilation");
+  llvm::SmallDenseMap<const Job *, std::unique_ptr<llvm::Timer>, 16>
+    DriverTimers;
 
   // Set up a callback which will be called immediately after a task has
   // started. This callback may be used to provide output indicating that the
   // task began.
-  auto taskBegan = [this] (ProcessId Pid, void *Context) {
+  auto taskBegan = [&] (ProcessId Pid, void *Context) {
     // TODO: properly handle task began.
     const Job *BeganCmd = (const Job *)Context;
 
+    if (ShowDriverTimeCompilation) {
+      llvm::SmallString<128> CmdLine;
+      {
+        llvm::raw_svector_ostream OS(CmdLine);
+        BeganCmd->printCommandLine(OS);
+      }
+      DriverTimers.insert({
+        BeganCmd,
+        std::unique_ptr<llvm::Timer>(
+          new llvm::Timer(CmdLine.str(), DriverTimerGroup))
```

Comment:
Bonus tip: if you call `str()` on the raw_svector_ostream instead of the SmallString, you don't have to force its destructor to be called.